smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/AuthSchemeResolverGenerator.kt

```diff
@@ -0,0 +1,291 @@
+                    }
+                }
+                // Handle default case, where operations default to auth schemes defined on service shape
+                val validSchemesForService =
+                    serviceIndex.getEffectiveAuthSchemes(ctx.service, ServiceIndex.AuthSchemeMode.NO_AUTH_AWARE)
+                write("default:")
+                renderSwitchCase(validSchemesForService, writer)
+            }
+        }
+    }
+
+    private fun renderSwitchCase(schemes: Map<ShapeId, Trait>, writer: SwiftWriter) {
+        writer.apply {
+            indent()
+            schemes.forEach {
+                if (it.key == SigV4Trait.ID) {
+                    renderSigV4AuthOption(it, writer)
+                } else {
+                    write("validAuthOptions.append(AuthOption(schemeID: \"${it.key}\"))")
+                }
+            }
+            dedent()
+        }
+    }
+
+    private fun renderSigV4AuthOption(scheme: Map.Entry<ShapeId, Trait>, writer: SwiftWriter) {
+        writer.apply {
+            write("var sigV4Option = AuthOption(schemeID: \"${scheme.key}\")")
+            write("sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: \"${(scheme.value as SigV4Trait).name}\")")
+            openBlock("guard let region = serviceParams.region else {", "}") {
+                write("throw ClientError.authError(\"Missing region in auth scheme parameters for SigV4 auth scheme.\")")
+            }
+            write("sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: region)")
+            write("validAuthOptions.append(sigV4Option)")
+        }
+    }
+
+    private fun renderConstructParametersMethod(
+        hasSigV4: Boolean,
+        ctx: ProtocolGenerator.GenerationContext,
+        writer: SwiftWriter
+    ) {
+        writer.apply {
+            openBlock(
+                "public func constructParameters(context: HttpContext) throws -> \$L {",
+                "}",
+                ClientRuntimeTypes.Auth.AuthSchemeResolverParams
+            ) {
+                if (usesRulesBasedAuthResolver(ctx)) {
+                    write("return try Default${getSdkId(ctx) + AUTH_SCHEME_RESOLVER}().constructParameters(context: context)")
+                } else {
+                    openBlock("guard let opName = context.getOperation() else {", "}") {
+                        write("throw ClientError.dataNotFound(\"Operation name not configured in middleware context for auth scheme resolver params construction.\")")
+                    }
+                    if (hasSigV4) {
+                        write("let opRegion = context.getRegion()")
+                        write("return ${getSdkId(ctx) + ClientRuntimeTypes.Auth.AuthSchemeResolverParams.name}(operation: opName, region: opRegion)")
+                    } else {
+                        write("return ${getSdkId(ctx) + ClientRuntimeTypes.Auth.AuthSchemeResolverParams.name}(operation: opName)")
```

Comment:
Pull the logic out of the string template

Reply:
Factored out the common logic into a constant.